### ISSUE TITLE
[Feature Requiest] *scratch* clients

### ISSUE THREAD

**Issue by andreyorst** [mawww/kakoune#3267]
# TL;DR
My proposal is to make it possible to mark clients as `scratch` clients, that should die automatically after last non-scratch client dies, and are capable of displaying only `*scratch*` and `*fifo*` buffers (with or without `*debug*` modifier). Regular clients should remain the same, but the server should ignore scratch clients while checking for save states of remaining file-visiting buffers in order to kill regular client.

## In more details
Right now we have file-visiting buffers, and non-file-visiting buffers, such as `*scratch*` and `*fifo*`. There are some tools and plugins that create such buffers, like `grep` or [kaktree](https://github.com/andreyorst/kaktree). Kakoune tools and external plugins usually respect `toolsclient` option or create their own client options to make tight integration of some buffer with some client. For example, `kaktree`, when used in tmux, creates tmux pane, of specific size, then creates hooks that actually make this client show only one buffer, which is `kaktree` buffer. This client is not suited to display regular buffers, since it is `kaktreeclient`. The same could be said for `toolsclient` since we usually expect tools inside it. And such clients usually are not needed after we close Kakoune.

But there's a problem. Nothing prevents Kakoune from displaying another buffer in such clients. This problem extremely annoying when exiting Kakoune. 

For example imagine that we have:
- we have two clients: `client0` and `kaktree`,
- two oped file-visiting buffers `file_a` and `file_b[+]`, where
  - `file_a` is being displayed in `client0` client,
  - `file_b` is not displayed and has modified state,
- and `*kaktree*[scratch]` buffer displayed in `kaktree` client, which has hooks to ensure that only `*kaktree*` buffer is displayed in it.

We're closing `client0` and Kakoune checks if there are clients left, sees that there's `kaktree` client and happily closes `client0`. Now we have to close `kaktree` but in order to do it, we have to save or discard changes in `file_b` buffer. Switching to this buffer is not an option, because of hooks in `kaktree` client. So we have to create new client, or save without knowing what actually we're saving, or discard with `:q!`. We also can remove hooks and make this client regular one, but I believe the whole thing should not happen in the first place.

Emacs has something similar, as [dedicated window](https://www.gnu.org/software/emacs/manual/html_node/elisp/Dedicated-Windows.html). Windows in Emacs have pretty much the same meaning as clients in Kakoune terminology. But it just doesn't allow to display anything but certain buffer.

Scratch clients on the other hand, are as `*scratch*` buffers - intended to be loosed. If we kill scratch client, we're killing it without any checks, If we're killing regular client, we're checking if there's a client that can display unsaved buffer, which is not scratch client. When we're successfully killing last non-scratch client we're killing all scratch clients as well.

This topic was discussed on Kakoune Discourse ([link](https://discuss.kakoune.com/t/best-solution-to-autoclose-tools-clients-when-all-non-tools-exited/671)) before and @Screwtapello actually suggested the idea.

**Comment by eraserhd** [comment]
To me, I wonder if the fundamental problem here is that the kaktree client is intended to behave differently from other clients, and maybe that points to the idea that it should be something else? A separate curses program, for example, or at least separate Kakoune session?

Granted, the nice thing about making UIs out of editor scratch buffers is that 1. it eliminates a lot of coding, and 2. it is easier to interact with the editor.

**Comment by andreyorst** [comment]
kaktree was just an example. The same happens to me with all toolsclient
related facilities. I think toolsclient itself points to the idea that
client is not intended to behave as editor window.

Having separate kind of clients will make it possible for  Kakoune to have
many Emacs-like packages that utilize Kakoune as a renderer and input
hanler. Looking at Emacs I think this is not the worst tactic out of all.
Also Arom editor does the same thing if we look at Electron as a renderer.

Making tight integration between editor and another cli application is
nice, but the big point here is that user will not be able to modify
application to make it work how desired. In contrast Emacs' packages can be
edited directly which also one point why there's so many of those for any
purpose.
